django/db/models/sql/query.py

```diff
@@ -1372,12 +1372,19 @@ def build_filter(
         if hasattr(filter_expr, "resolve_expression"):
             if not getattr(filter_expr, "conditional", False):
                 raise TypeError("Cannot filter against a non-conditional expression.")
+            pre_joins = self.alias_refcount.copy()
             condition = filter_expr.resolve_expression(
-                self, allow_joins=allow_joins, summarize=summarize
+                self, allow_joins=allow_joins, reuse=can_reuse, summarize=summarize
             )
             if not isinstance(condition, Lookup):
                 condition = self.build_lookup(["exact"], condition, True)
-            return WhereNode([condition], connector=AND), []
+            used_joins = {
+                k for k, v in self.alias_refcount.items() if v > pre_joins.get(k, 0)
+            }
+            return (
+                WhereNode([condition], connector=AND),
+                used_joins if not current_negated else (),
```

Comment:
The `require_outer` logic below is more complex below so I think we'll want to add move coverage around usage of things like `IsNull(expr, True)` as well. 

This logic is not flawless though anyway (ticket-25946)